### PR TITLE
Bug 1796348,  Bug 1811795 - Target Android 13 in Fenix and AC

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/main/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1811795
https://bugzilla.mozilla.org/show_bug.cgi?id=1796348

### CODE REVIEW

**Comment by rocketsroger** [comment]
Adding a "do not land" label so we can have a plan before landing this.  Crash reporter depends on the crash notification to report crashes.  If we update the API to 13 too early, we might lose a lot of crash reports on Android 13 devices.

There are two things we can do:
1. Make notification pre permission prompt on by default (Currently it's in v110 as an experiment)
2. Add a "Looks like Firefox crashes" prompt at start if there are unreported crashes to prompt the user to report their crashes.

I think the minimum here is to do step 1 first before landing this change.  Thanks

**Comment by mcarare** [comment]
> Adding a "do not land" label so we can have a plan before landing this. Crash reporter depends on the crash notification to report crashes. If we update the API to 13 too early, we might lose a lot of crash reports on Android 13 devices.
> 

Why would we lose crash reports on Android 13 devices? This patch introduces no changes in the behavior. We actually make sure that every time we try to show a notification from anywhere in the app we ask for the notifications permission. We do not have control over that in the current state.  And we do not have callbacks for the result of our request made with `ensureMarketingChannelExists` exists calls. I do not think we should postpone this any longer for changes that can be introduced after targeting Android 13 lands. 

If anything, this allows showing any kind of prompt/ dialog/before asking the notifications permission from anywhere in the app, using the request permission method in the delegate. 

Also, for crash reports, we can have the alternative of showing a dialog, if the notification permission is denied. Because the logic can be handled within the delegate. So we can prompt for crashes we are not able to send now for Android 13 devices

**Comment by rocketsroger** [comment]
> 



> > Adding a "do not land" label so we can have a plan before landing this. Crash reporter depends on the crash notification to report crashes. If we update the API to 13 too early, we might lose a lot of crash reports on Android 13 devices.
> 
> IMO this patch cannot cause losing crash reports on Android 13 devices. This patch introduces no changes in the behavior. We actually make sure that every time we try to show a notification from anywhere in the app we ask for the notifications permission. We do not have control over that in the current state. And we do not have callbacks for the result of our request made with `ensureMarketingChannelExists` exists calls. I do not think we should postpone this any longer for changes that can be introduced after targeting Android 13 lands.

Yes, but we can only ask once.  Currently there is an experiment of using a dialog showing why it's in the user's advantage to give us the notification permission and then ask for the permission.  This is a better way to ensure we have a higher opt-in for notification permission.

> If anything, this allows showing any kind of prompt/ dialog/before asking the notifications permission from anywhere in the app, using the request permission method in the delegate.
> 
> Also, for crash reports, we can have the alternative of showing a dialog, if the notification permission is denied. Because the logic can be handled within the delegate. So we can prompt for crashes we are not able to send now for Android 13 devices

This doesn't work consistently so we had to move to a notification.  Showing dialog when Fenix is not in foreground doesn't work consistently on newer OS versions.

**Comment by mcarare** [comment]
> Yes, but we can only ask once. Currently there is an experiment of using a dialog showing why it's in the user's advantage to give us the notification permission and then ask for the permission. This is a better way to ensure we have a higher opt-in for notification permission.
> 

I agree. But this patch does not prevent us from using that dialog. Also, with this patch, you can ask for permission twice.

File: fenix/app/src/main/java/org/mozilla/fenix/onboarding/HomeNotificationPermissionDialogFragment.kt
Comment (mcarare) [inline fenix/app/src/main/java/org/mozilla/fenix/onboarding/HomeNotificationPermissionDialogFragment.kt:56]: Depending on the experiment results we could now better handle a potential rejection and show an additional message (even by just changing the text) before we ask for the permission a second time.

**Comment by pwd-github** [comment]
The notification conversation is obviously scope creep. 

The thing is, most people click yes on everything. The ones that click no, need clear instructions as to how to turn notifications back on. If the reason is to submit crashes, that's a terrible reason.

But none of that has anything to do with this patch, especially given that all competitor browsers are targeting Android 13 in their release versions.

**Comment by rocketsroger** [comment]
I've confirmed that we do have pre-permission notification on by default starting v112.  This is one of the requirement to unblock this PR.  

I am confirming with data-science to have the result on the pre-permission prompt experiment to make sure that the result is positive before we land this.

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt
Comment (rahulsainani) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt:65]: Think we can remove this as we will target 33 and above and this will always be true

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt
Comment (rahulsainani) [inline android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt:81]: ⚠️ This could throw a `RemoteException`.  

As it's already being checked in `notificationsDelegate.requestNotificationPermission()` maybe we can call that already without the conditions.

File: fenix/app/src/main/java/org/mozilla/fenix/onboarding/HomeNotificationPermissionDialogFragment.kt
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/onboarding/HomeNotificationPermissionDialogFragment.kt:53]: Do we need this now? cc @rocketsroger

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt
Comment (mcarare) [reply android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt]: Done.

File: android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt
Comment (mcarare) [reply android-components/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/manager/FetchDownloadManager.kt:78]: Removed the additional check.

**Review by rahulsainani — APPROVED** [review]
Works well! Thanks for persevering and working through all the different feedback cycles 🚀

**Review by rocketsroger — APPROVED** [review]
Removed my objection and the `do not land` label.  🚢

File: android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt
Comment (jonalmeida) [inline android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt:39]: Why hold an instance to the exception here? 🤔

File: android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt
Comment (jonalmeida) [inline android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt:31]: nit: doc for non-existent property.

File: android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt
Comment (jonalmeida) [inline android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt:42]: Discussion only: for our apps right now, we hold a singleton reference to a NotificationDelegate so we it's a bit harder to leak here. What do you think about wrapping this in a weakref (as a follow-up)?

```suggestion
    private val notificationPermissionHandler: MutableMap<WeakReference<AppCompatActivity>, ActivityResultLauncher<String>> =
```

File: android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt
Comment (jonalmeida) [inline android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt:66]: Since we call this in `onDestroy`, I'm more inclined to having a weak ref wrapper since we know that `onDestroy` isn't always called.

File: android-components/components/tooling/lint/src/main/java/mozilla/components/tooling/lint/NotificationManagerChecks.kt
Comment (jonalmeida) [inline android-components/components/tooling/lint/src/main/java/mozilla/components/tooling/lint/NotificationManagerChecks.kt:31]: Love this! 🙌

File: android-components/components/tooling/lint/src/main/java/mozilla/components/tooling/lint/NotificationManagerChecks.kt
Comment (jonalmeida) [inline android-components/components/tooling/lint/src/main/java/mozilla/components/tooling/lint/NotificationManagerChecks.kt:66]: Curious, why are these two checks separate? Wouldn't it be the same to do:

```kotlin
if (inComponentPackage || inAppPackage) {
    context.report(
        ISSUE_NOTIFICATION_USAGE,
        node,
        context.getLocation(node),
        NOTIFY_ERROR_MESSAGE,
    )
}
```

**Review by jonalmeida — APPROVED** [review]
Excellent work - this PR was much easier to read than I expected!

I've only reviewed a few core components in this change since it has already been reviewed by two others. All my comments can be addressed after this lands so it's non-blocking.

:shipit:

File: android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt
Comment (mcarare) [reply android-components/components/support/base/src/main/java/mozilla/components/support/base/android/NotificationsDelegate.kt:39]: It was left from a previous iteration, but I will change it to create an instance only when thrown, as it is a single place where we do that.

File: android-components/components/tooling/lint/src/main/java/mozilla/components/tooling/lint/NotificationManagerChecks.kt
Comment (mcarare) [reply android-components/components/tooling/lint/src/main/java/mozilla/components/tooling/lint/NotificationManagerChecks.kt:66]: I encountered an issue when having it in the single if (or) statment would miss some valid lint issues. I am going to leave it like this and check-up on the issue as a follow-up.

**Pull request merged**
